aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -143,50 +143,82 @@ static void check_1d(const Tensor& t, const char* arg, const char* fn) {
 }
 
 Tensor addr(const Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
-  check_1d(vec1, "vec1", "addr");
-  check_1d(vec2, "vec2", "addr");
-  Tensor b_self;
-  std::tie(b_self) = expand_size(self, {vec1.size(0), vec2.size(0)}, "addr");
-  return at::_addr(b_self, vec1, vec2, beta, alpha);
+  Tensor result = at::empty({0}, self.options());
+  at::addr_out(result, self, vec1, vec2, beta, alpha);
+  return result;
 }
 
 Tensor& addr_(Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
-  check_1d(vec1, "vec1", "addr");
-  check_1d(vec2, "vec2", "addr");
-  return at::_addr_(self, vec1, vec2, beta, alpha);
+  return at::addr_out(self, self, vec1, vec2, beta, alpha);
 }
 
 Tensor& addr_out(Tensor &result, const Tensor& self, const Tensor& vec1, const Tensor& vec2, Scalar beta, Scalar alpha) {
   check_1d(vec1, "vec1", "addr");
   check_1d(vec2, "vec2", "addr");
-  Tensor b_self;
-  std::tie(b_self) = expand_size(self, {vec1.size(0), vec2.size(0)}, "addr_out");
-  return at::_addr_out(result, b_self, vec1, vec2, beta, alpha);
+
+  Tensor self_;
+  if (&result != &self) {
+    std::tie(self_) = expand_size(self, {vec1.size(0), vec2.size(0)}, "addr_out");
+    result.resize_as_(self_);
+    result.copy_(self_);
+  } else {
+    self_ = self;
+  }
+
+  TORCH_CHECK(self_.dim() == 2,
+              "2D tensor expected, got ", self_.dim(), "D tensor for input");
+  TORCH_CHECK(self_.size(0) == vec1.size(0) && self_.size(1) == vec2.size(0),
+              "size mismatch",
+              ", input: ", self_.sizes(),
+              ", v1: ", vec1.sizes(),
+              ", v2: ", vec2.sizes());
+
+  // If beta is 0, self would be ignored
+  // and `nan` and `inf` in it would not be propagated
+  if (beta.to<double>() == 0.0) {
+    result.zero_();
+  } else if (beta.to<double>() != 1.0) {
+    result.mul_(beta);
+  }
+
+  // addr is implemented as a composite op through torch.outer
+  at::add_out(result,
+              result,
+              at::outer(vec1, vec2),
+              alpha);
+  return result;
 }
 
 Tensor& ger_out(Tensor &result, const Tensor& self, const Tensor& vec2) {
-  check_1d(self, "self", "ger");
-  check_1d(vec2, "vec2", "ger");
-  if (result.dim() != 2 || result.size(0) != self.size(0) || result.size(1) != vec2.size(0)) {
-    result.resize_({ self.size(0), vec2.size(0) });
-  }
-  // resize_ does the "broadcasting", don't need to broadcast again.
-  return at::_addr_out(result, result, self, vec2, Scalar(0), Scalar(1));
+  TORCH_WARN("torch.ger is deprecated and may be removed, "
+             "please use torch.outer instead.");
+  return at::outer_out(result, self, vec2);
 }
 
+// torch.ger, alias for torch.outer
 Tensor ger(const Tensor& self, const Tensor& vec2) {
   Tensor result = at::empty({0}, self.options());
```

Comment:
This won't work with type promotion. Just do:

```
return self.outer(vec2); 
```

Reply:
Thanks for the kind suggestion, the code has been updated.